Created "AnimatorParameterBinding" to add support for binding to Animator parameters

MorganMoon opened the pull request:
- Created AnimatorParameterBinding
Works similar to OneWayPropertyBinding:
![image](https://user-images.githubusercontent.com/10552305/33501146-1e5c0588-d6a1-11e7-9d17-a9adfdef66b3.png)
Works with adapters:
![image](https://user-images.githubusercontent.com/10552305/33501172-354761b6-d6a1-11e7-9969-a6cbdbe88acc.png)

The dropdown for "View property" includes the parameters & their type of the Animator on the same GameObject:
![image](https://user-images.githubusercontent.com/10552305/33501261-9ba091f8-d6a1-11e7-8505-755f9f475095.png)


- Created an inspector editor class for the AnimatorParameterBinding (AnimatorParameterBindingEditor)
- Added new type "AnimatedParameterTrigger" in order to bind to the Animator paramater type "Trigger"
Example of use:
![image](https://user-images.githubusercontent.com/10552305/33501033-ae21f548-d6a0-11e7-8c6b-14cca0880521.png)

They work just like triggers on the Animator work, you can call Set() or Reset() on these properties, and it implements INotifyPropertyChanged itself, so you don't need to worry about that for these properties


The AnimatorParameterBinding is required to be on the same GameObject as the Animator it is binding to. 😄

Commit:
Created "AnimatorParameterBinding" to add support for binding to Animator parameters.

- Created AnimatorParameterBinding
- Created an inspector editor class for the AnimatorParameterBinding (AnimatorParameterBindingEditor)
- Added new type "AnimatedParameterTrigger" in order to bind to the Animator paramater type "Trigger"

Commit:
Removed uneeded Assert in ToggleActiveBinding

RoryDungan commented on UnityWeld/Binding/AnimatorParameterBinding.cs:
```diff
@@ -0,0 +1,227 @@
+        /// <summary>
+        /// Animator to use
+        /// </summary>
+        private Animator _animator;
```
Convention in the rest of the project is to just use lowerCamelCase for private fields. I know that there is already a property inherited from MonoBehaviour called `animator` but I'd probably just either use `private new Animator animator;` to explicitly override the inherited member, or just give it a different name like maybe `boundAnimator`.

RoryDungan commented on UnityWeld/Binding/AnimatorParameterBinding.cs:
```diff
@@ -0,0 +1,227 @@
+            switch (AnimatorParameterType)
+            {
+                case AnimatorControllerParameterType.Float:
+                    {
```
I think you can safely remove the parentheses in this switch statement. In the rest of the project we usually only bother using parentheses in `case` statements in cases where we are definining new variables - see here: https://stackoverflow.com/questions/3652408/c-sharp-switch-statement-with-without-curly-brackets-whats-the-difference

RoryDungan commented on UnityWeld/Binding/Internal/PropertyEndPoint.cs:
```diff
@@ -49,6 +49,19 @@ public PropertyEndPoint(object propertyOwner, string propertyName, IAdapter adap
+                type = this.propertyOwner.GetType();
+                this.property = type.GetProperty(this.propertyName);
+            }
+
```
I don't think you actually have to modify PropertyEndPoint to get it to work with AnimatorParameterTrigger. propertyName and propertyOwner are both passed into this constructor anyway so we shouldn't be overriding them for specific types of properties. 

Having a single method that does different things in specific circumstances is generally bad as it makes it less obvious what the method does from looking at the interface, and also makes it more error prone due to more code paths. Is there a way we can incorporate this part of the functionality into AnimatorParameterTrigger instead?

RoryDungan commented on UnityWeld/Binding/ToggleActiveBinding.cs:
```diff
@@ -70,11 +70,6 @@ public override void Connect()
-                "ToggleActiveBinding can only be bound to a boolean property."
-            );
-
             var propertySync = new PropertySync(
```
Removing this assert is a good catch now that ToggleActiveBinding supports adapters :+1:

RoryDungan commented on UnityWeld_Editor/AnimatorParameterBindingEditor.cs:
```diff
@@ -0,0 +1,337 @@
+            // So we disable this all while in play mode
+            if (EditorApplication.isPlaying)
+            {
+                EditorGUILayout.HelpBox("Cannot modify in play mode!", MessageType.Warning);
```
Maybe we should disable the rest of the controls as well to make sure that users don't try to modify it while in play mode.

MorganMoon replied:
What do you mean? I should add that to all of the binding editors?

RoryDungan replied:
Hmm I guess this would apply to all the editors. If it's a problem with the others then probably don't worry about adding this, but I'll add a task to the roadmap Trello that we should either make it so you *can* edit in play mode, or all of the editors are disabled in play mode.

RoryDungan commented on UnityWeld_Editor/AnimatorParameterBindingEditor.cs:
```diff
@@ -0,0 +1,337 @@
+        {
+            var animator = targetScript.GetComponent<Animator>();
+
+            AnimatorControllerParameter[] properties = new AnimatorControllerParameter[] { };
```
This doesn't need to be initialised to `new AnimatorControllerParameter[] { }` because it is assigned to again before it is ever used.

RoryDungan commented on UnityWeld_Editor/AnimatorParameterBindingEditor.cs:
```diff
@@ -0,0 +1,337 @@
+            // Animator is on is inactive causes it to return an empty array, aswell as fill
+            // the unity console with warnings. This behaviour is undocumented, and this is just
+            // a work around which seems to work fine
+            if (!animator.gameObject.activeInHierarchy)
```
Maybe the code for getting the properties out of an animator could be refacored out to a separate method. It seems like it could be easily separated from the logic for showing the parameters menu and the method is very large already.

RoryDungan replied:
you did what the previous comment was talking about so I'm not sure why it didn't remove that comment, but thanks :+1:

RoryDungan commented on UnityWeld_Editor/AnimatorParameterBindingEditor.cs:
```diff
@@ -0,0 +1,337 @@
+            // a work around which seems to work fine
+            if (!animator.gameObject.activeInHierarchy)
+            {
+                UnityEditorInternal.ComponentUtility.CopyComponent(animator.transform);
```
I'm very hesitant to use anything that calls to `UnityEditorInternal` because it is an undocumented API. This means that any Unity release, even one marked as a minor or bugfix one, could potentially remove it, change the way it works breaking our code, or add unintended side effects.

MorganMoon replied:
I will look more into what else I could do. I think just getting the scale, position, and rotation and the setting those again will achieve the same thing :)

RoryDungan replied:
Cool, thanks 👍

RoryDungan commented on UnityWeld_Editor/AnimatorParameterBindingEditor.cs:
```diff
@@ -0,0 +1,337 @@
+                    var isActive = animator.gameObject.activeSelf;
+                    animator.gameObject.SetActive(false);
+                    animator.gameObject.SetActive(true);
+                    animator.gameObject.SetActive(isActive);
```
Toggling the object active does seem like an odd fix. I think it would be worth expanding the comment a bit to explain why it is necessary/what breaks if we don't do it this way.

MorganMoon replied:
Yeah, not a great comment sorry. I forgot exactly what this fixes, which just shows the comment needs to be better lol. I'll look into why I am doing this

RoryDungan commented on UnityWeld_Editor/AnimatorParameterBindingEditor.cs:
```diff
@@ -0,0 +1,337 @@
+                    "Set Animator parameter"
+                );
+
+                switch (newSelectedProperty.type)
```
See commend above about style for switch statements. I also noticed this is an exact duplicate of the code at line 244 - would be good if this was refactored out into a method that takes an `AnimatorControllerParameterType` and returns a `Type` to get rid of the duplication.

RoryDungan commented on UnityWeld_Editor/AnimatorParameterBindingEditor.cs:
```diff
@@ -0,0 +1,337 @@
+                    return;
+                }
+
+                switch (properties[selectedIndex].type)
```
see above

RoryDungan commented on UnityWeld_Editor/AnimatorParameterBindingEditor.cs:
```diff
@@ -0,0 +1,337 @@
+                        viewModelPropertyPrefabModified = property.prefabOverride;
+                        break;
+
+                    case "uiPropertyName":
```
There is no "uiPropertyName" field on AnimatorParameterBinding so this is unnecessary. You should add cases for `animatorParameterName` and `animatorParameterType` though.

Actually seeing this makes me realise that since I renamed a bunch of fields on the other binding classes, I'll need to update this function in the other custom editors too. Thanks!

RoryDungan commented on UnityWeld_Editor/AnimatorParameterBindingEditor.cs:
```diff
@@ -0,0 +1,337 @@
+
+            public override int GetHashCode()
+            {
+                return new { Name = Name, Type = Type }.GetHashCode();
```
I think this can be simplified to `return new { Name, Type }.GetHashCode()`

RoryDungan reviewed (changes requested):
Hey! Sorry it's taken me so long to get to this, I've just been a bit busy. 

Overall I think this is a very useful addition to the library so thanks for submitting it! There are some changes that would need to be made before I think we can merge it though so have a look at the comments I've left in the code for details. Feel free to respond to my comments if there's anything you want more details on or don't agree with.

RoryDungan commented:
I also noticed that adding the component to a new object that previously didn't have an Animator on it spams this warning every frame:
```Animator is not playing an AnimatorController
UnityEngine.Animator:get_parameters()
UnityWeld_Editor.AnimatorParameterBindingEditor:ShowAnimatorParametersMenu(GUIContent, Action`1, AnimatorParameterTypeAndName, Type&) (at Assets/Unity-Weld/Editor/AnimatorParameterBindingEditor.cs:180)
UnityWeld_Editor.AnimatorParameterBindingEditor:OnInspectorGUI() (at Assets/Unity-Weld/Editor/AnimatorParameterBindingEditor.cs:60)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)
```

We should fix this before merging too.

RoryDungan commented:
I also noticed that adding the component to a new object that previously didn't have an Animator on it spams this warning every frame:
```Animator is not playing an AnimatorController
UnityEngine.Animator:get_parameters()
UnityWeld_Editor.AnimatorParameterBindingEditor:ShowAnimatorParametersMenu(GUIContent, Action`1, AnimatorParameterTypeAndName, Type&) (at Assets/Unity-Weld/Editor/AnimatorParameterBindingEditor.cs:180)
UnityWeld_Editor.AnimatorParameterBindingEditor:OnInspectorGUI() (at Assets/Unity-Weld/Editor/AnimatorParameterBindingEditor.cs:60)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)
```

We should fix this before merging too.

Commit:
Work on requested changes:
- Changed AnimatorParameterBinding._animator to boundAnimator
- Changed switch style
- Removed PropertyEndPoint changes and not just use the constructor correctly
- Broke apart methods in AnimatorParameterBindingEditor
- Fixed spammy unity messages

Commit:
Merge branch 'master' of https://github.com/Real-Serious-Games/Unity-Weld into feature/AnimatorParameterBinding

Commit:
Anonymous type no longer uses named properties

Commit:
Made suggested fixes
- Expanded upon a vague comment
- stopped using UnityEditorInternal
- All binding editors are disabled during play mode in the editor

MorganMoon commented:
Alright, all requested changes should be there :)

RoryDungan commented on UnityWeld_Editor/AnimatorParameterBindingEditor.cs:
```diff
@@ -0,0 +1,343 @@
+                case AnimatorControllerParameterType.Trigger:
+                        return typeof(AnimatorParameterTrigger);
+                default:
+                        return null;
```
Maybe instead of returning null this could throw an ArgumentOutOfRangeException. You have accounted for all the types of AnimatorControllerParameterTypes that exist in Unity at the moment but it is still good to account for the possibility that in the future they may introduce another one. If that ever happened, I think having an exception thrown here with an appropriate message would make more sense than just returning null, which could cause a NullReferenceException somewhere else unexpected in the code.

RoryDungan reviewed (commented):
Thanks for making those changes! There's just one other thing that occurred to me is probably worth changing so I've added a comment but overall looks really good. 

Me and the rest of the RSG dev team are on holidays at the moment so I'll test it and hopefully merge once we're all back next week.

Commit:
Changed return null to an ArgumentOutOfRangeException in AnimatorParameterBindingEditor.AnimatorControllerParameterTypeToType()

RoryDungan committed:
Changed prompt for modifying bindings in play mode to just an info

RoryDungan committed:
Implemented IEquatable<T> for AnimatorParameterTypeAndName

RoryDungan committed:
Simplified switch statement, adjusted formatting and made method static

RoryDungan committed:
Made unsupported animator parameter type throw exception rather than return empty string
